Enable p-refinement via an Event

kidder opened the pull request:
## Proposed changes

This PR will allow p-refinement to be done as an Event.
This requires the DG scheme to reactively check if the communicated mortar data is on the correct Mesh.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Use forward declarations in NeighborMesh

kidder committed:
Add missing communication of integration order

This was missed in PR 6157.

kidder committed:
Communicate boundary_correction_mesh in BoundaryData

This is the mesh corresponding to boundary_correction_data.  It needs
to be communicated in case the mortar needs to be reprojected after
AMR increases resolution.

kidder committed:
Rename variable to match tag name

kidder committed:
Improve error message in MortarHelpers

kidder committed:
Always communicate volume DG mesh in BoundaryData

This is what will be used by reactive p-refinement for DG.

kidder committed:
Add event to change grid resolution

kidder committed:
Project mortar data if resolution changed

In order to reactively adapt to p-refinement done as an Event, the mortar
data may need to be projected to a new mortar mesh.

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp:
```diff
@@ -23,10 +23,13 @@ Mesh<Dim> mortar_mesh(const Mesh<Dim>& face_mesh1,
+    ASSERT(face_mesh1.basis(i) == Spectral::Basis::Legendre and
+               face_mesh2.basis(i) == Spectral::Basis::Legendre,
+           "Only Legendre basis meshes are supported for element faces so far, "
+           "but "
```
Combine lines

nilsdeppe commented on src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp:
```diff
@@ -0,0 +1,162 @@
+class RefineMesh : public Event {
+ public:
+  /// \cond
+  explicit RefineMesh(CkMigrateMessage* /*unused*/) {}
```
move to cpp

nilsdeppe commented on src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp:
```diff
@@ -0,0 +1,162 @@
+  using options = tmpl::list<>;
+  static constexpr Options::String help = {"Perform p-refinement"};
+
+  RefineMesh() = default;
```
move to cpp

nilsdeppe commented on src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp:
```diff
@@ -0,0 +1,162 @@
+  static constexpr Options::String help = {"Perform p-refinement"};
+
+  RefineMesh() = default;
+
```
Do you need a `pup` function?

nilsdeppe commented on src/ParallelAlgorithms/Amr/Events/RefineMesh.hpp:
```diff
@@ -0,0 +1,162 @@
+        if (decision[d] != amr::Flag::Split and
+            decision[d] != amr::Flag::Join) {
+          overall_decision[d] = std::max(overall_decision[d], decision[d]);
+        }
```
Should we choose p-refine if we get split? And p-coarsen if we get Join? Or just ignore for now? It might be good to have an ERROR since this would indicate a decision is being ignored.

wthrowe commented on src/Evolution/DiscontinuousGalerkin/BoundaryData.hpp:
```diff
@@ -20,6 +20,7 @@ namespace evolution::dg {
  * 2. the volume mesh corresponding to the ghost cell data. This allows eliding
  *    projection when all neighboring elements are doing DG.
+ * 3. the mortar mesh of the data on the mortar
  * 3. the mesh of the neighboring element's face (not the mortar mesh!)
```
Renumber the later entries.  (The numbering being wrong actually confuses doxygen, which surprises me.)

wthrowe commented on src/Evolution/DiscontinuousGalerkin/InboxTags.hpp:
```diff
@@ -44,6 +44,7 @@ namespace evolution::dg::Tags {
  * 2. the volume mesh corresponding to the ghost cell data. This allows eliding
  *    projection when all neighboring elements are doing DG.
+ * 3. the mortar mesh of the data on the mortar
  * 3. the mesh of the neighboring element's face (not the mortar mesh!)
```
Again, fix numbering.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp:
```diff
@@ -23,10 +23,13 @@ Mesh<Dim> mortar_mesh(const Mesh<Dim>& face_mesh1,
+               face_mesh2.basis(i) == Spectral::Basis::Legendre,
+           "Only Legendre basis meshes are supported for element faces so far, "
+           "but "
+           "face_mesh1 is "
```
[optional] formatting

wthrowe commented on src/ParallelAlgorithms/Amr/Events/Events.hpp:
```diff
@@ -0,0 +1,10 @@
+
+/// \ingroup AmrGroup
+namespace amr {
+/// \brief %Eventss for adaptive mesh refinement
```
Sspelling.

wthrowe commented on src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp:
```diff
@@ -457,6 +485,28 @@ bool receive_boundary_data_local_time_stepping(
+              boundary_data_history->at(mortar_id).local().for_each(
+                  project_boundary_mortar_data);
+              boundary_data_history->at(mortar_id).remote().for_each(
+                  project_boundary_mortar_data);
```
Call `boundary_data_history->at(mortar_id).clear_coupling_cache()`.

wthrowe commented on src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp:
```diff
@@ -472,13 +522,19 @@ bool receive_boundary_data_local_time_stepping(
+            if (mortar_mesh !=
+                received_mortar_data->second.boundary_correction_mesh) {
+              boundary_data_history->at(mortar_id).remote().for_each(
+                  project_boundary_mortar_data);
```
Again, clear the cache, unless this is only modifying the entry that was just inserted, in which case add a comment that you don't have to clear the cache.

wthrowe commented on src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp:
```diff
@@ -779,9 +836,25 @@ struct ApplyBoundaryCorrections {
+                         << "\n mortar_mesh: " << mortar_mesh << "\n");
+              ASSERT(local_data_on_mortar.number_of_grid_points() ==
+                         neighbor_data_on_mortar.number_of_grid_points(),
+                     "RIP\n");
```
Better message?  Probably print out the values.

nilsdeppe replied:
A bit more info would be helpful :)

wthrowe commented on src/ParallelAlgorithms/Amr/Events/Events.hpp:
```diff
@@ -4,7 +4,5 @@
 /// \brief %Eventss for adaptive mesh refinement
-namespace Events {}
-}  // namespace amr
+namespace amr::Events {}  // namespace amr::Events
```
Move reformatting to the commit where it was added.  Also remove the same-line closing comment.

wthrowe commented on tests/Unit/Evolution/DiscontinuousGalerkin/Actions/Test_ApplyBoundaryCorrections.cpp:
```diff
@@ -684,14 +684,14 @@ void test_impl(const Spectral::Quadrature quadrature,
-                                            {flux_data},
-                                            {neighbor_next_time_step_id},
-                                            decision};
+      const evolution::dg::BoundaryData<Dim> data{mesh,
```
Move to appropriate commit.

wthrowe reviewed (changes requested):
clang-tidy errors are not in our code and probably wrong, so nothing to do for those.

knelli2 commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp:
```diff
@@ -93,12 +94,14 @@ void p_project(
     const gsl::not_null<MortarDataHistoryType*> mortar_data_history,
     const Mesh<Dim>& new_mesh, const Element<Dim>& new_element,
     const std::unordered_map<ElementId<Dim>, amr::Info<Dim>>& neighbor_info,
+    const ::dg::MortarMap<Dim, Mesh<Dim>>& neighbor_mesh,
```
Should these changes go in a different commit?

knelli2 reviewed (commented):
No need to wait for my approval

kidder committed:
Do not communicate interface_mesh in BoundaryData

This is the appropriate slice (which can be obtained from the
mortar_id) of the volume_mesh

nilsdeppe reviewed (dismissed):
My requests have been addressed. I don't need to look at this again before it's merged.

wthrowe commented:
Still missing the LTS cache management.  Have you tested this in an LTS evolution?

kidder committed:
Use new Event RefineMesh in EvolveScalarWave

wthrowe commented:
We discussed some methods of testing this elsewhere.  I assume you've done some of those?

wthrowe commented:
Discussed elsewhere.  The tests have been done.
